polygeist/lib/Dialect/Polygeist/Transforms/ArgumentPromotion.cpp

```diff
@@ -0,0 +1,491 @@
+    assert(!isRecursiveCall(callOp) && "Callee must not be recursive");
+  }
+
+  /// Returns true if the operation \p op is a callable operation with private
+  /// visibility and a defined callable region, and false otherwise.
+  static bool isPrivateCallable(Operation *op);
+
+  /// Returns true if the call \p callOp is recursive, and false otherwise.
+  static bool isRecursiveCall(CallOpInterface callOp);
+
+  /// Returns true if \p type is a pointer to a struct (memref<? x struct<>>),
+  /// and false otherwise.
+  static bool isValidMemRefType(Type type);
+
+  /// Perform the transformation.
+  void transform();
+
+private:
+  /// Identify call operands that should be peeled and peel them, returns the
+  /// number of struct members peeled. Note: this function populates a map
+  /// between the original operand and the peeled values.
+  unsigned peelMembers();
+
+  /// Modify the call by replacing the original operands with their
+  /// corresponding peeled members.
+  void modifyCall();
+
+  /// Modify the called function by replacing the original operands with their
+  /// corresponding peeled members.
+  void modifyCallee();
+
+  /// Replace the argument at position \p pos in the region \p callableRgn with
+  /// \p newArgs. The reference parameter \p newArgAttrs is filled with the new
+  /// argument attributes.
+  void replaceArgumentWith(unsigned pos, Region &callableRgn,
+                           const SmallVector<Value> &newArgs,
+                           SmallVector<Attribute> &newArgAttrs) const;
+
+  /// Replace uses of the argument \p origArg in the region \p callableRgn with
+  /// the arguments starting at position \p pos in the callable region.
+  void replaceUsesOfArgument(Value origArg, unsigned pos,
+                             Region &callableRgn) const;
+
+  /// The candidate call site.
+  CallOpInterface callOp;
+
+  /// Map between the position of the original call operand and its
+  /// corresponding peeled arguments.
+  std::map<unsigned, SmallVector<Value>> operandToMembersPeeled;
+};
+
+class ArgumentPromotionPass
+    : public polygeist::impl::ArgumentPromotionBase<ArgumentPromotionPass> {
+public:
+  using ArgumentPromotionBase<ArgumentPromotionPass>::ArgumentPromotionBase;
+
+  void runOnOperation() override;
+
+private:
+  /// Populate \p candidates with call operations to transform.
+  void collectCandidates(SmallVectorImpl<Candidate> &candidates);
+
+  /// Return true if the call operation \p callOp is a candidate and false
+  /// otherwise.
+  bool isCandidate(CallOpInterface callOp);
+
+  /// Return true is the calee is a candidate and false otherwise.
+  bool isCandidate(CallableOpInterface callableOp);
+};
+
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// Candidate
+//===----------------------------------------------------------------------===//
+
+bool Candidate::isPrivateCallable(Operation *op) {
```

Comment:
Should `isPrivateCallable` take a `CallableOpInterface`?
It currently returns true e.g., on an add operation.